### ISSUE TITLE
How to limit cross correlation window width in Numpy?

### ISSUE THREAD

**Issue by bringingheavendown** [scipy/scipy#4940]
I am learning numpy/scipy, coming from a MATLAB background. The xcorr function in Matlab has an optional argument "maxlag" that limits the lag range from –maxlag to maxlag. This is very useful if you are looking at the cross-correlation between two very long time series but are only interested in the correlation within a certain time range. The performance increases are enormous considering that cross-correlation is incredibly expensive to compute.

In numpy/scipy it seems there are several options for computing cross-correlation. numpy.correlate, numpy.convolve, scipy.signal.fftconvolve. If someone wishes to explain the difference between these, I'd be happy to hear, but mainly what is troubling me is that none of them have a maxlag feature. This means that even if I only want to see correlations between two time series with lags between -100 and +100 ms, for example, it will still calculate the correlation for every lag between -20000 and +20000 ms (which is the length of the time series). This gives a 200x performance hit! Is it possible that I can contribute this feature?

**Comment by jdmonaco** [comment]
I'm just a scipy user (not a dev), but I definitely agree that this has been a issue in many of my analyses. My usual solution is to just roll my own functions that implement maxlag limits. For example, to cross-correlate spike-train data, I use [something like this](https://gist.github.com/jdmonaco/f0ecd87a9f8428ff030f#file-xcorr_maxlag-py).

A similar sliding window approach could be used for continuous time series. However, that kind of data (dense, as opposed to spike timing) would probably benefit from a lower-level implementation using something like `scipy.weave`.

**Comment by sturlamolden** [comment]
Just a tip, recode your spike-train data to a series of 0s and 1s by assuming your spike-train is a sum of delta functions (look up Koch and Segev, "Methods in Neuronal Modelling", an old book but still just as relevant). Then use rfft to compute the cross-correlation. That is usually faster than doing those summations, Also, if you use a fast FFT library (MKL, FFTW, Apple Accelerate) it will also be multi-threaded and fast as hell. :)

**Comment by josef-pkt** [comment]
It would be interesting to see some performance comparisons between fft and a simple loop for a few correlation/lags, to see where the threshold is.

**Comment by jdmonaco** [comment]
At the risk of veering OT since @bringingheavendown is probably considering continuous data, there are problems with the conventional FFT method for spiking data. The necessary binarization (@sturlamolden's "recoding") can only ever approximate a summation of delta functions, so either you use more memory for your binary vectors (and correspondingly longer FFT wall time) as you try to combat binning noise with smaller bins or you simply deal with increasingly inaccurate correlations. Binning noise is compounded by the FFT convolution since it operates across the entire duration of the recording data and becomes a real problem when your bins are too large. In contrast, the windowed loop implementation is based on exact (well, floating point) subtraction of timestamps (i.e., precise to 1/32,000 second) which at least gives you accurate lag counts in the resulting correlograms. If we consider the number of bins in the resultant correlogram as a parameter that the user of the cross-correlation function should be able to control, then using the FFT method also requires rebinning to the specified number of bins and range of lags based on the maxlag parameter, and this can introduce even more aliasing noise. 

I did a quick comparison (Mid-2012 Retina Macbook Pro, 2.6Ghz) and found that computing the 1-second maxlag autocorrelation of a random 10 Hz spike train of 200-second duration takes about 0.35 seconds for the loop implementation I posted above compared to about 19 seconds for an FFT convolution with 0.0005-second binarization and correlogram rebinning. Even with these tiny binarization bins, you can see a fair amount of variance in the noise floor of the FFT result relative to the loop result: 
![xcorrs-0 5ms](https://cloud.githubusercontent.com/assets/370169/8026587/beccfd38-0d4a-11e5-81ac-3cade0097109.png)

To get the two methods to run in about the same wall time, I increased the FFT bin size to 1 ms, producing the correlation in about 0.48 seconds. However, the variance in the noise floor increased substantially (including an apparent upward bias, which I'm not sure about): 
![xcorrs-1ms](https://cloud.githubusercontent.com/assets/370169/8026586/b4b74664-0d4a-11e5-873c-f156e829d87c.png)

[Here's the code for the comparison.](https://gist.github.com/db1f34fe993dffa71d26) This could be extended to sweep across (e.g.) firing rate to see the crossing point for the two methods, as suggested by @josef-pkt. Regardless, I've used an `IPython.parallel` version of the looped algorithm to compute autocorrelations of spike trains with 100,000s of spikes across several hours of recording in just a few seconds. This performance is pretty ok with me. There are different considerations for continuous data, which are already natively sampled, but windowed algorithms can be more efficient and accurate at least for point-processes like spiking or other sparse data sets. It would be nice see functionality addressing this in Scipy, but it could get confusing given all the correlate/convolve/fft functions already there.

**Comment by sturlamolden** [comment]
You need to low-pass filter before you bin the spike train, for the same reason you have an anti-alias filter before the ADC.

**Comment by jdmonaco** [comment]
Ah you could've mentioned that in your simple FFT recipe

**Comment by jdmonaco** [comment]
If data has been collected it is already after the ADC, so filtering/decimation just loses precision again in the name of efficiency. In the end, this all depends on the goals of the particular data and analysis.

**Comment by sturlamolden** [comment]
Well, this is getting totally OT :) Anyway, a spike-train coded as a sum of delta functions is an analog signal. So when you bin the signal into rate or count histograms you will introduce aliasing, and your binning noise is the aliasing. The theory is explained in the classical French  and Holden (1971) paper, though it is uncommon to use their "exact" anti-alias filter for reasons of efficiency. The easiest solution is to convolve each delta-function spike with a short FIR low-pass filter before you sample the spike train. http://link.springer.com/article/10.1007/BF00291117

**Comment by jdmonaco** [comment]
I take all this to mean that a `maxlag` parameter should only be implemented the way it is in `matplotlib.Axes.xcorr` which is just `np.correlate(..., mode=2)` followed by slicing out the maxlag window. I poked around in Matlab and can't find any evidence that the `maxlag` argument affects the length of computation anyway, only the size of the results. Thanks for the reference. :)

**Comment by sturlamolden** [comment]
It depends on the implementation in Matlab. But having a maxlag should improve the performance. The complexity is O(n*n) for a cross-correlation of two signals of length n. With maxlag of m it should be O(n*m), which can be much smaller.

**Comment by sturlamolden** [comment]
Travis failure on Pyhton 2.6 is because I forgot to conditionally define ```NPY_UNLIKELY()```.

**Comment by jdmonaco** [comment]
Right, I just thought you were saying there was no point to implementing windowed correlation. The reduction in complexity is the entire reason for this issue I think, so that is understood. It would be nice to have a core C code implementation available as an `np.correlate`/`np.convolve` mode or similar, but not sure how critical that would be given the various workarounds.

**Comment by sturlamolden** [comment]
I think NumPy code is supposed to be "lean and mean", so anything with extra features belongs more naturally in SciPy.

**Comment by bringingheavendown** [comment]
Here is my concrete proposal:
1) Add maxlags as an additional argument to numpy.correlate.  This will automatically add the functionality to numpy.convolve, as it calls numpy.correlate for the heavy lifting.  
2) Change scipy.signal.correlate and scipy.signal.convolve so that they call numpy.correlate for the heavy lifting because there is no reason to have two independent implementations of the exact same function.  
After this is done, performance analysis will be done to determine the vector lengths at which correlate with maxlags is faster than fftconvolve, and that information will be added to the documentation.  
The other similar functions (scipy.signal.convolve2d, scipy.signal.correlate2d, scipy.ndimage.filters.convolve, scipy.ndimage.filters.convolve1d, scipy.ndimage.filters.correlate, scipy.ndimage.filters.correlate1d) will have to wait for some other time as they are not immediately interchangeable with numpy.correlate.  
Can you all let me know how this sounds and what the process would be for going about this?
Thank you.

**Comment by jaimefrio** [comment]
On Mon, Jun 8, 2015 at 3:55 PM, bringingheavendown <notifications@github.com
> wrote:

> Here is my concrete proposal:
> 1) Add maxlags as an additional argument to numpy.correlate. This will
> automatically add the functionality to numpy.convolve, as it calls
> numpy.correlate for the heavy lifting.
>
Most numpy devs read this mailing list too, but if you want to change
numpy, you will need to discuss it in the numpy list as well.

> 2) Change scipy.signal.correlate and scipy.signal.convolve so that they
> call numpy.correlate for the heavy lifting because there is no reason to
> have two independent implementations of the exact same function.
>
I believe signal's convolve and correlate functions do multidimensional
convolution/correlation, while numpy only does 1-D, so they are really not
interchangeable.


> After this is done, performance analysis will be done to determine the
> vector lengths at which correlate with maxlags is faster than fftconvolve,
> and that information will be added to the documentation.
>
> The other similar functions (scipy.signal.convolve2d,
> scipy.signal.correlate2d, scipy.ndimage.filters.convolve,
> scipy.ndimage.filters.convolve1d, scipy.ndimage.filters.correlate,
> scipy.ndimage.filters.correlate1d) will have to wait for some other time as
> they are not immediately interchangeable with numpy.correlate.
>
> Can you all let me know how this sounds and what the process would be for
> going about this?
>
Send a pull request to numpy, preferably after discussing the details over
there, they are always welcome...

Jaime

-- 
(\__/)
( O.o)
( > <) Este es Conejo. Copia a Conejo en tu firma y ayúdale en sus planes
de dominación mundial.

**Comment by sturlamolden** [comment]
As @jaimefrio said, they are not identical. Another thing is that scipy.ndimage should perhaps be left alone because of GSoC.

**Comment by bringingheavendown** [comment]
By the way, Matlab's xcorr function uses the fft trick, so it's maxlag parameter doesn't save much time.  Matplotlib.pyplot.axes.xcorr also only uses maxlag for sizing the output, but doesn't actually change the computation at all.  

I implemented my own [xcorr_maxlag function](https://gist.github.com/bringingheavendown/b4ce18aa007118e4e084) for continuous data, and even in straight python, it gives a 50x speedup on my data compared to numpy.correlate (although it is still slightly slower than fftconvolve).  I am using maxlag 10, time series length 21000.  numpy.correlate takes 0.336 s, my xcorr_maxlag takes 0.006 s and fftconvolve takes 0.005 s.  

I'm assuming that x and y are the same length here.  I'm not sure what the appropriate way to deal with different sized vectors is.

**Comment by sturlamolden** [comment]
Different sized arrays is what you e.g. have in a matched filter.

**Comment by bhargavvader** [comment]
Would any of you guys be able to explain the difference between the three functions OP talked abouy (numpy.correlate, numpy.convolve, scipy.signal.fftconvolve)?

Also, why is xcorr talked in the same breath as correlate and convolute? Don't they give fairly different results? 
Excuse my ignorance in this, I've only been dabbling in DSP for about a month or so.

**Comment by sturlamolden** [comment]
> Also, why is xcorr talked in the same breath as correlate and convolute? Don't they give fairly different > results? 

Cross-correlation is what you use to implement a *matched filter*. It is basically convolution without time-reversal of the filter kernel. A cross-correlogram in statistical signal analysis is a particular case of matched filtering. But it is not the one use case.

**Comment by sturlamolden** [comment]
> Would any of you guys be able to explain the difference between the three functions OP talked about (numpy.correlate, numpy.convolve, scipy.signal.fftconvolve)?

```numpy.correlate``` does a cross-correlation of a one-dimensional signal and a one-dimensional  template.

```numpy.convolve``` does a convolution of a one-dimensional signal and a one-dimensional kernel.

```scipy.signal.correlate``` does a cross-correlation of an N-dimensional signal and a N-dimensional  template.

```scipy.signal.convolve``` does a convolution of an N-dimensional signal and an N-dimensional kernel.

```scipy.signal.fftconvolve``` is the same as ```scipy.signal.convolve``` but computes the convolution using FFT.

**Comment by bhargavvader** [comment]
@sturlamolden , thank you!

**Comment by bringingheavendown** [comment]
I have now implemented this functionality in numpy.correlate() and numpy.convolve(). https://github.com/bringingheavendown/numpy. The files that were edited are:
numpy/core/src/multiarray/multiarraymodule.c
numpy/core/numeric.py
numpy/core/tests/test_numeric.py
Please look over the code, my design decisions, and the unit tests I have written. This is my first time contributing, so I am not confident about any of these and welcome feedback.

**Comment by sturlamolden** [comment]
Which branch? (Don't work in your master branch. I once did, and its stupid.)

If you post a WIP PR it is easier to compare the changes to the current NumPy.

**Comment by sturlamolden** [comment]
No // comments in ANSI C.

**Comment by sturlamolden** [comment]
I have a feeling that what you really want is a **function that computes cross-correlograms**, the way they are used in statistical signal analysis. You are messing with functions which are actually about filtering, for which this "maxlag" and "lagstep" thing makes no sence, nor the "lagvec" output. WTF is a "maxlag" or a "lagstep" in a convolution anyway?

(Not to mention that inplace modification is not the way we return optional output arrays.)

It would be more appropriate to implement a function in ```scipy.signal``` that returns a cross-correlogram. It should return two arrays, correlations and time lags, and should allow multiple forms of normalisations.

**Comment by sturlamolden** [comment]
If you want to plot a cross-correlogram, ```matplotlib.pyplot.xcorr``` will do what you want. It also returns the correlations and the time-lags. A similar function in SciPy should not do any plotting. 

This would not be the first case of overlap between SciPy and matlplotlib, though. There is another one between ```scipy.signal.welch``` and ```matplotlib.pyplot.psd```.

http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.xcorr

**Comment by bringingheavendown** [comment]
I do want a function that computes cross-correlograms, however the implementation is exactly the same for cross-correlograms as convolution.  Not only that, is the numpy.correlate() function not for computing cross-correlograms?  Maxlag and lagstep still make sense in the context of convolution.  Say you have a time series (this is not the best example) of rain amounts and you have a kernel for plant growth given rain in the recent past.  Your time series is the entire year, but you are only interested in the plant growth during the months of April through August.  Not only that, you do not need daily readout of plant growth; weekly resolution is enough for your needs.  You wouldn’t want to compute the convolution for the entire time series, instead you would do: numpy.convolve(rain, growth_kernel, (april, september, 7), lagvec) and get lagvec back with the indices of the sundays in april through august, and a return vector with the amount of plant growth on those days.  

I am happy to change the return of lagvec; I wasn’t sure how to return optional output arrays.  It didn’t seem that returning both was appropriate because if you called the function with a single output argument, you would still get a tuple of arrays, which you would then have to unpack.  Please let me know what would be better.  

My intention is to eventually also implement normalization as an option in this function.  Perhaps also the fft shortcut to computation of correlation/convolution as another option.  Right now, the maxlag option is what I am working on.  

I don’t really want to add an entirely new function to scipy.signal.  It was already hard enough as a new user trying to figure out which of the five seemingly identical functions in numpy, scipy, and matplotlib that I should be using.  Besides, if all of these functions are essentially doing the same computation, there should only be a single base implementation that they all use, so that 1) the learning curve is decreased and 2) that any optimizations be passed on to all of the functions instead of having to be independently reimplemented several times.  So, even if we do decide that scipy.signal should have a new correlogram command, it should be a wrapper for numpy.correlate.  But why wouldn't one just use scipy.signal.correlate for the 1d case as well?

As for matplotlib.pyplot.xcorr, it is simply a wrapper for numpy.correlate.  As I described at the start of this conversation, the maxlag functionality gives great performance improvements, but matplotlib.pyplot.xcorr's maxlag option doesn’t actually give this performance improvement.  It actually calculates the entire cross-correlogram and then only returns part of it. 

In conclusion, I do think that this enhancement should be added to numpy.correlate.  I do hope that you give suggestions as to how to improve the interface in terms of parameters and outputs.

**Comment by bringingheavendown** [comment]
Here are the specific issues that could use attention in this implementation:

How to deal with the optional lagvec output array?
Is it ok that this implementation breaks the previous undocumented behavior of mode=0, 1, or 2 returning the correlation for mode='valid', 'same', or 'full' respectively?  If not, how should maxlag input be passed?
Should the checks that are included in convolve() be added to correlate()?
Is it ok that I used INT_MAX to represent requests for a default argument in the C code?
Do I need to check for "default arguments" at all, given that the calling functions should now *always* call _pyarray_correlate with the appropriate number of arguments?  I kind of feel there is some legacy behavior I should be maintaining, but I'm not 100% sure.  
Is the error message in _pyarray_correlate "PyErr_SetString(PyExc_ValueError, "mode must be 0, 1, 2, or 3")" still valuable?  What should it be replaced with?
Are there other error checks I should add to the code at this or other points?
Could you all clean up the unit tests/add some new ones to be absolutely sure that it deals properly with weird combinations of input array sizes and lag inputs?  I couldn't figure out how to run all the same unit tests with the order of input arrays switched.

**Comment by sturlamolden** [comment]
You would use an optional input parameter which indicates whether to return lagvec. It should default to False. If it is True you return a tuple of two arrays ibstead of just one array.

**Comment by sturlamolden** [comment]
INT_MAX is not ok because we have 64 bit systems. You need to use an invalid value.

**Comment by bringingheavendown** [comment]
Why does 64 bit systems matter for using INT_MAX?  What invalid value would I use?  (int)NULL=0 and (int)void throws an error.  

I used INT_MAX as the default values of minlag, maxlag, and lagstep in array_correlate2 in multiarraymodule.c, which is the function called directly by python.  Those are currently optional arguments, but both convolve() and correlate() do actually pass those.  INT_MAX gets passed through functions until it gets to _pyarray_correlate in multiarraymodule.c, which is the function that actually implements the cross-correlation.  It checks for INT_MAX values, but it's kind of awkward because if we got that far without calculating minlag etc. then something else is wrong.

**Comment by sturlamolden** [comment]
Because INT_MAX is the maximum value of an int, which accirding to the ANSI C standard is at least 0x7FFF, usually 0x7FFFFFFF on contemporary systems. On a 64 bit system the maximum value of an npy_intp is 0x7FFFFFFFFFFFFFFF. Since npy_intp is what we use to store an array size, you must use an invalid value for a maxlag instead of something that can be as small as 0x7FFF. If negative values of maxlag is disallowed (I haven't cared to look) you can e.g. use -1.

**Comment by bringingheavendown** [comment]
Unfortunately, negative values of lags are allowed: for example, if you want to look at the correlation only for lags for which the second vector is to the left of the first vector.

**Comment by sturlamolden** [comment]
Then you must e.g. increase the number of arguments to the C function and have one with boolean bitflags or whatever.

**Comment by bringingheavendown** [comment]
Amazing!  Thank you!

**Comment by sturlamolden** [comment]
The size of an array of 0x7FFFFFFF C double is only about 16 GB. There are people running NumPy on much larger computers than this.

**Comment by sturlamolden** [comment]
You could use a silly value like ```NPY_MAX_INTP```. It would never overflow with an array of double. Or to be pedantic you can just disallow this integer value before calling C.

**Comment by bringingheavendown** [comment]
The problem is that I can't figure out how to get bitflags to work with PyArg_ParseTupleAndKeywords.  I have to pass in a variable that it either modifies or doesn't modify.  I still have to initialize the variable with some value that I check to see if PyArg_ParseTupleAndKeywords changed.  I'll go with NPY_MAX_INTP instead of INT_MAX.

**Comment by sturlamolden** [comment]
I am not sure I understood. You cannot figure out how to pass an integer from Python to C?

**Comment by bringingheavendown** [comment]
I can't figure out how to check in the C code if an integer was passed from the python caller without using a default value.  Eg.
```
    int mode = 0, maxlag = NPY_MAX_INTP, minlag = NPY_MAX_INTP, lagstep = NPY_MAX_INTP;
    static char *kwlist[] = {"a", "v", "mode", "minlag", "maxlag", "lagstep", NULL};

    if (!PyArg_ParseTupleAndKeywords(args, kwds, "OO|iiii", kwlist,
                &a0, &shape, &mode, &minlag, &maxlag, &lagstep)) {
        return NULL;
    }
```

**Comment by sturlamolden** [comment]
```C
int mode = 0, maxlag = NPY_MAX_INTP, minlag = NPY_MAX_INTP, lagstep = NPY_MAX_INTP;
```
Integer overflow.

**Comment by bringingheavendown** [comment]
so i want 
```
npy_intp  mode = 0, maxlag = NPY_MAX_INTP, minlag = NPY_MAX_INTP, lagstep = NPY_MAX_INTP;
```
?

**Comment by sturlamolden** [comment]
```C
    int mode = 0;
    npy_intp maxlag = NPY_MAX_INTP, minlag = NPY_MAX_INTP, lagstep = NPY_MAX_INTP;
    static char *kwlist[] = {"a", "v", "mode", "minlag", "maxlag", "lagstep", NULL};
    if (!PyArg_ParseTupleAndKeywords(args, kwds, "OO|innn", kwlist,
                &a0, &shape, &mode, &minlag, &maxlag, &lagstep)) {
        return NULL;
    }
```

(But I still think you should avoid using default values for maxlag like this.)

**Comment by bringingheavendown** [comment]
so what do you recommend?

**Comment by sturlamolden** [comment]
I am not sure.

**Comment by sturlamolden** [comment]
```C

    #define HAS_MAXLAG 1
    #define HAS_MINLAG 2
    #define HAS_LAGSTEP 4

    int mode = 0;
    int flags=0;
    npy_intp maxlag, minlag, lagstep;
    PyObject *tmp;
    static char *kwlist[] = {"a", "v", "mode", "minlag", "maxlag", "lagstep", NULL};

    tmp = PyDict_GetItemString(kwds, "maxlag");
    if (tmp != NULL) flags |= HAS_MAXLAG;
    Py_XDECREF(tmp);

    tmp = PyDict_GetItemString(kwds, "minlag");
    if (tmp != NULL) flags |= HAS_MINLAG;
    Py_XDECREF(tmp);

    tmp = PyDict_GetItemString(kwds, "lagstep");
    if (tmp != NULL) flags |= HAS_LAGSTEP;
    Py_XDECREF(tmp);

    if (!PyArg_ParseTupleAndKeywords(args, kwds, "OO|innn", kwlist,
                &a0, &shape, &mode, &minlag, &maxlag, &lagstep)) {
        return NULL;
    }
```

**Comment by bringingheavendown** [comment]
where do the #define's go?

**Comment by sturlamolden** [comment]
Just ignore it. You have a layer of Python between the C function and the user. Set the fag there and parse it as an int.

**Comment by bringingheavendown** [comment]
That is already taken care of.  The python code that I have written only ever calls this C function with values included for maxlag etc.  What I was trying to deal with is the possibility that at some point someone else writes a python function in numpy that doesn't use these optional arguments when calling the C function.  Should I just assume that they will do their due diligence and not do that?

**Comment by sturlamolden** [comment]
Then just give them default values which retains the current behaviour and do not break backwards compatibility.

What would that be? Perhaps -NPY_MAX_INTP for minlag and NPY_MAX_INTP for maxlag? (I haven't checked your code, so I'm not sure.)

**Comment by bringingheavendown** [comment]
I don't quite understand what you are saying.  Is it possible to talk on skype or something similar? Is that against protocol here?

**Comment by endolith** [comment]
will this be implemented in scipy.fftconvolve, too?  Would it just be truncation of the output in that case?  Eventually we'd like to include fftconvolve functionality inside convolve/correlate:  https://github.com/scipy/scipy/issues/2651

**Comment by bringingheavendown** [comment]
lag modes would have to be independently implemented in scipy.fftconvolve.  My original intention was to include the fft option in numpy.correlate, but it has taken too long to get the lag mode into numpy.correlate altogether that i don’t really have time any more.  lag modes could either truncate fft convolve output or could be used as an alternative.  performance tradeoffs between using lag modes and fft depend on the lengths of the vectors being convolved and the number of lags of interest.

**Comment by endolith** [comment]
yes, the goal of https://github.com/scipy/scipy/issues/2651 would be to have a `method=` with `'auto'` as default that would choose whichever implementation works best for the input types and sizes.

**Comment by rgommers** [comment]
Since this discussion is related to ``scipy.signal`` functionality let's keep it open, but current discussion/review is ongoing in gh-5978. Once that PR is merged, let's come back to this one - we'd like the API addition in the exact same format as will be added to numpy I'd think.

**Comment by atabakd** [comment]
Just wanted to +1 this.

**Comment by sturlamolden** [comment]
This is also related to a stalled NumPy PR, for some reason also numbered 5978.

https://github.com/numpy/numpy/pull/5978

**Comment by HoniSanders** [comment]
Hi all, 
For those of you still interested, I have put a new PR on numpy for this functionality ([#31261](https://github.com/numpy/numpy/pull/31261)).  I started a thread on the scipy discussion board as suggested by the numpy maintainers (https://discuss.scientific-python.org/t/custom-lag-parameters-on-correlate-and-convolve-functions/2307).  Feel free to share any thoughts you have there or here.

**Comment by HoniSanders** [comment]
Would people be willing to comment on the scipy discussion board for this topic? 
 (https://discuss.scientific-python.org/t/custom-lag-parameters-on-correlate-and-convolve-functions/2307)

Thank you
